### PR TITLE
Adding files for VectorPenaltyDirichletBC #15673

### LINKED ISSUE #15673: Adding VectorPenaltyDirichletBC to MOOSE framework
## Reason

This feature is needed to allow for Vector variables to be used with the `PenaltyDirichletBC` object and could be useful for implementing the penalty dirichlet BC when using the `navier_stokes` module since the velocity can be stored as a vector.
## Design

Moved the `VectorPenaltyDirichletBC` test object from the MooseTestApp into the MooseApp framework in the `framework/src/bcs` folder and `framework/include/bcs` folder. Added markdown for the BC in the `framework/doc/content/source/bcs/` folder. The BC accepts a nonlinear vector variable and performs the penalty Dirichlet boundary condition routine to weakly enforce the boundary to a user-specified value using a user-defined scalar penalty value. 
## Impact

This BC will allow for vector-type nonlinear variables to be used with the `PenaltyDirichletBC` and increase the amount of types available to use the `PenaltyDirichletBCs` within the MOOSE system.

### PR DESCRIPTION
**Pull request by amherm**
Added the VectorPenaltyDirichletBC files from the test system into the moose framework to all apps to have access to it. The test for this can be found in test/tests/vector_fe/electromagnetic_coulomb_gauge.i. Closes #15673

### CODE REVIEW

File: framework/src/bcs/VectorPenaltyDirichletBC.C [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "VectorPenaltyDirichletBC.h"
+#include "Function.h"
+
+registerMooseObject("MooseTestApp", VectorPenaltyDirichletBC);
```
Comment (loganharbour) [inline framework/src/bcs/VectorPenaltyDirichletBC.C:13]: The object is still registered to the test app. Change this by:

```suggestion
registerMooseObject("MooseApp", VectorPenaltyDirichletBC);
```

**Review by loganharbour — CHANGES_REQUESTED** [review]

**Comment by loganharbour** [comment]
Wait... you didn't move files here, you added them again.

Remove the newly added files and move the ones from the test dir:
```
git rm framework/src/bcs/VectorPenaltyDirichletBC.C framework/include/bcs/VectorPenaltyDirichletBC.h 
git mv test/src/bcs/VectorPenaltyDirichletBC.C framework/src/bcs/VectorPenaltyDirichletBC.C
git mv test/include/bcs/VectorPenaltyDirichletBC.h framework/include/bcs/VectorPenaltyDirichletBC.h
```
...and then you should be good to make the object registration change.

File: framework/src/bcs/VectorPenaltyDirichletBC.C (hunk 1 above)
Comment (amherm) [reply framework/src/bcs/VectorPenaltyDirichletBC.C]: I think I fixed this now and amended my commit with a force-push

**Comment by loganharbour** [comment]
Unfortunately not: https://civet.inl.gov/job/573949/

You need to add a brief class description via `params.addClassDescription()` the parameter declaration for the class. This is something that isn't required for test objects but is for standard objects/

File: framework/doc/content/source/bcs/VectorPenaltyDirichletBC.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+!syntax description /BCs/VectorPenaltyDirichletBC
+
+## Description
+
+`VectorPenaltyDirichletBC` implements a vector form of [`PenaltyDirichletBC`](/PenaltyDirichletBC.md),
+which enforces a weak sense of the Dirichlet boundary
+condition by setting the boundary as a penalty times the inner product of the
+test function and the difference between the current solution and Dirichlet
+data. This is given by $p(\\vec{u}^\\ast, \\vec{u} - \\vec{u}_0)$, where p is a
```
Comment (zachmprince) [inline framework/doc/content/source/bcs/VectorPenaltyDirichletBC.md:11]:
```suggestion
data. This is given by $p(\vec{u}^\ast, \vec{u} - \vec{u}_0)$, where $p$ is a
```

File: framework/doc/content/source/bcs/VectorPenaltyDirichletBC.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+Dirichlet data.
+
+This boundary condition can be useful for problems where the mesh is not as
+refined and could potentially smooth out the problem data on a coarser mesh. It,
+however, has problems on refined meshes and leads to an ill-conditioned problem,
+which can be difficult to solve.
+
+## Example Input Syntax
+
+!listing test/tests/kernels/vector_fe/electromagnetic_coulomb_gauge.i
```
Comment (zachmprince) [inline framework/doc/content/source/bcs/VectorPenaltyDirichletBC.md:23]:
```suggestion
!listing test/tests/kernels/vector_fe/electromagnetic_coulomb_gauge.i block=BCs
```

File: framework/src/bcs/VectorPenaltyDirichletBC.C [hunk 4]
```diff
@@ -21,7 +21,10 @@ VectorPenaltyDirichletBC::validParams()
   params.addParam<FunctionName>("y_exact_sln", 0, "The exact solution for the y component");
   params.addParam<FunctionName>("z_exact_sln", 0, "The exact solution for the z component");
   params.addParam<bool>("linear", false, "Whether this is formulated as a linear problem");
-  return params;
+  params.addClassDescription("Enforces a Dirichlet boundary condition for "
+                             "vector nonlinear variables in a weak sense by "
+                             "applying a penalty to the difference in the "
+                             "current solution and the Dirichlet data.") return params;
```
Comment (zachmprince) [inline framework/src/bcs/VectorPenaltyDirichletBC.C:27]:
```suggestion
  params.addClassDescription("Enforces a Dirichlet boundary condition for "
                             "vector nonlinear variables in a weak sense by "
                             "applying a penalty to the difference in the "
                             "current solution and the Dirichlet data.");
  return params;
```

**Review by zachmprince — DISMISSED** [review]
Small changes.

**Comment by loganharbour** [comment]
BTW for future reference: if you go to "files changed", you can batch those "suggested changes" into one commit. We'll do a squash on merge.

You should have also been added a team in the idaholab group so your tests will auto run on any of your new PRs.

**Review by loganharbour — APPROVED** [review]
👍🏼 welcome to the moose contributors club

**Pull request merged**